### ISSUE TITLE
K-Means IP has increasing objective, but better performance - logging issue? 

### ISSUE THREAD

**Issue by GerardMaggiolino** [facebookresearch/faiss#1937]
# Summary

When running k-means with `spherical=True`, final classification results are improved compared to using an L2 distance metric when the features are unit normed. This is expected. 

However, when inspecting training loss with with `.obj` attribute, the loss increases with each iteration. I'm not sure what's causing this discrepancy. As I'm using k-means++ by initializing the centroids manually with `nredo=1` and selecting the best of multiple runs, the `.obj` attribute needs to be accurate to select the lowest loss model. 

# Platform

Running on:
- [X] CPU
- [ ] GPU

Interface: 
- [ ] C++
- [X] Python

# Reproduction instructions

Run any unit normed dataset and inspect the `.obj` attribute with `spherical=True`. It will be increasing per iteration, although the final model will perform well.

**Comment by GerardMaggiolino** [comment]
Alternatively, is it possible to supply multiple runs to `init_centroids` to the `.train()` function to have a set of centroids per iteration of `nredo`?

**Comment by mdouze** [comment]
The objective is the sum of "distances" returned by the clustering index.
For an IP index the distances are actually dot products, that are better when higher, so it makes sense that the objective is increasing. 
NB that spherical k-means and IP search there is no clear guarantee or loss that k-means optimizes. 

For the nredo / init_centroids: indeed it's a bit useless to use the combination of both.... A workaround is to run the optimization several times in an external loop.

**Comment by GerardMaggiolino** [comment]
@mdouze Thank you, this solves my question :)
